### PR TITLE
fix: Various Combobox fixes with tests.

### PR DESCRIPTION
**Pull request by atmgrifter00**
# Pull Request

## 📖 Description

This PR aims to address several issue that I recently filed against the Combobox. Most of the changes herein relate to appropriately updating the value of the Combobox itself and sending the proper change notification when that happens.

### 🎫 Issues

- #6142 
- #6111
- #6106

## 👩‍💻 Reviewer Notes

Probably the most interesting thing that's changing in this PR that may not seem totally obvious to the listed issues is the changing from calling `updateValue` (changes Combobox value) in certain places (like when pressing the `Tab` key or arrow keys while the popup is open) to only updating the content/value of the `input`. This, to me, accomplishes a couple of things: 1) allows the Combobox to more easily report actual changes (emit the `change` event) when the Combobox value differs from the `input` value, and 2) provides an easier understanding of when the Combobox _should_ emit a `change` event (when we update the Combobox `value`). Admittedly, I didn't take this all the way as I left a couple places that were calling `updateValue` without emitting the event alone: when the form resets, and when `slottedOptionsChanged`. This may merit discussion.

## 📑 Test Plan

I have written unit tests for the each of the issues mentioned above.

### General

- [x] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [x] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [inline packages/web-components/fast-foundation/src/combobox/combobox.ts:596]: I made this new private method so that we maintain current behavior for things like pressing 'Tab', where it would just put the cursor at the end of the selected item from the popup (that was navigated to via arrow keys), and not actually select the text.

File: packages/web-components/fast-foundation/src/combobox/combobox.spec.ts
Comment (atmgrifter00) [inline packages/web-components/fast-foundation/src/combobox/combobox.spec.ts:227]: Hmmm...I don't think this is valid, as I'm not awaiting this call to `nextUpdate()`, and if I did, I would presume this could introduce intermittency to the test. I'll think about another way to write this test.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts]: Reverted this.

File: packages/web-components/fast-foundation/src/combobox/combobox.spec.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.spec.ts]: Simple re-org of code addressed this.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [inline packages/web-components/fast-foundation/src/combobox/combobox.ts:617]: Not sure how I feel about this method as it may not be appropriate to use it to make the determination to emit the `change` event (like in the case of a form reset). However, this seemed preferable to duplicating the logic in two places.

**Comment by atmgrifter00** [comment]
@chrisdholt Hey Chris. I was just curious if there was a particular timeframe I could expect this PR to be reviewed. Thanks!

**Comment by chrisdholt** [comment]
> Edit
> Hide
> Delete

@radium-v has been under the weather this week and I think his review is likely most important here as he has the _most_ context on any implications here. I'll try and take a look by EOW this week as well, but definitely want John to take a look before we merge.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (radium-v) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:405]: In `this.filterOptions()`, `this.filter` is set to `""` if `this.autocomplete` is undefined or `none`. That means that this would always reset the `selectedIndex` to -1 on input, even if the current value matches the text of an available option.

**Review by radium-v — CHANGES_REQUESTED** [review]

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:405]: Note that the `selectedIndex` will get updated appropriately upon commit. As is, I believe the behavior matches what is there currently (when `autocomplete` is set to `none`). However, I could see it being nice to update the `selectedIndex` to the appropriate item whilst typing (again, when `autocomplete` is `none`). Would you want me to do this, or would you rather just introduce a guard to prevent us setting `selectedIndex` to -1 if there is no autocomplete.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [inline packages/web-components/fast-foundation/src/combobox/combobox.ts:506]: @radium-v, I removed this because I don't think it really does anything. If anything, it seems to be unnecessary overhead, as we will filter twice for certain keystrokes like `Delete` and `Backspace`. I'm not sure why we would want to set `selectedIndex` to -1 for any of these key strokes. To me, the `inputHandler` is sufficient for doing the filtering and setting `selectedIndex` to the appropriate value.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [inline packages/web-components/fast-foundation/src/combobox/combobox.ts:669]: @radium-v,  I recognized another bug while testing changes. If `this.value` gets updated to the first option (`firstSelectedOption.text`), then `this.control.value` will not represent the current value, which is a problem. I've added a test to reflect this change.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:406]: @radium-v, as you can see, I am now setting `selectedIndex` to whatever the exact match is for the non-inline autocomplete cases. I've added tests to reflect this change.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts]: Updated as suggested.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:669]: Thinking about this a bit more, I'm not sure why `this.control.value` doesn't automatically get updated when `this.value` is set through the template binding. Thoughts @radium-v?

**Review by chrisdholt — APPROVED** [review]
Approved pending resolution and sign off by @radium-v per his requests. Thanks for your patience and the contribution!

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (chrisdholt) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:669]: @EisenbergEffect is it possible that this is a bug? `this.control` relates to a ref in the template, and `this.value` is observable using the binding syntax as such:

```ts
html`
    <template>
        <input :value="${x => x.value}" ${ref("control")}>
    </template>
`
```

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (radium-v) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:506]: This is needed for autocomplete `inline` and `both` functionality to work properly. When navigating through a filtered list after typing text, the auto-filled content is automatically highlighted. Pressing any of these keys will cause the input selection to change, which changes the autocomplete comparison value.

**Review by radium-v — CHANGES_REQUESTED** [review]

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:506]: Could we at least remove the Backspace and Delete cases (as these should be handled by the `inputHandler`)?

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:697]: Writing to the DOM happens on the update queue. So, if a binding updates value, you have to wait for the queue to be processed before the DOM is updated. Writing to models is synchronous but writing to the DOM is "eventually consistent". This is for performance reasons. That would be my guess at what's going on here based on the quick read of the above.

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (radium-v) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:410]: The two lines can be reduced to `if (e.inputType.includes("deleteContent")) {` which should cover all three `deleteContent` types

**Review by radium-v — APPROVED** [review]

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts:693]: I'll assume that I can just leave this change in place. (as I'm not sure what the alternative to correcting the bad behavior would be).

File: packages/web-components/fast-foundation/src/combobox/combobox.ts
Comment (atmgrifter00) [reply packages/web-components/fast-foundation/src/combobox/combobox.ts]: Updated.

**Pull request merged**
